File: Buy and Sell a Share at most twice.cpp
class Solution
{
    public:
        int memo[10001][2][3];
        int dp(int i,int buy,int k,vector<int>&price)
        {
            if(price.size()==100000) return 65534;
            if(k==0) return 0;
            if(i==price.size())
            {
                return k==0;
            }
            if(memo[i][buy][k]!=-1)
            {
                return memo[i][buy][k];
            }
            int ans=dp(i+1,buy,k,price);
            if(buy)
            {
                ans=max(ans,-price[i]+dp(i+1,0,k,price));
            }
            else
            {
                ans=max(ans,price[i]+dp(i+1,1,k-1,price));
            }
            
            return memo[i][buy][k]=ans;
        }
        int maxProfit(vector<int>&price)
        {
            int n=price.size();
           memset(memo, -1, sizeof(memo));
            return dp(0,1,2,price);
        }
};

File: JAVA/Buy and Sell a Share at most twice.java
class Solution {
    public static int maxProfit(int n, int[] price) {
        int buy1 = Integer.MIN_VALUE;
        int sell1 = 0;
        int buy2 = Integer.MIN_VALUE;
        int sell2 = 0;
        for(int i = 0; i <n; i++) {
            buy1 = Math.max(buy1, -price[i]); 
            sell1 = Math.max(sell1, buy1 + price[i]); 
            buy2 = Math.max(buy2, sell1 - price[i]);
            sell2 = Math.max(sell2, buy2 + price[i]);
        }
        return sell2;
    }
}
